central/signatureintegration/datastore/datastore_impl.go

```diff
@@ -149,6 +160,90 @@ func (d *datastoreImpl) verifyIntegrationIDDoesNotExist(ctx context.Context, id
 	return nil
 }
 
+func (d *datastoreImpl) verifyIntegrationIDIsNotInPolicy(ctx context.Context, id string) error {
+	policyCtx := sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.Policy)))
+
+	policies, err := d.policyStore.GetAllPolicies(policyCtx)
+	if err != nil {
+		return errors.Wrap(err, "retrieving all policies")
+	}
+
+	var policiesContainingID []string
+	for _, p := range policies {
+		if checkIfPolicyContainsID(id, p) {
+			policiesContainingID = append(policiesContainingID, p.GetName())
+		}
+	}
+
+	if len(policiesContainingID) != 0 {
+		integration, err := d.getSignatureIntegrationByID(ctx, id)
+		if err != nil {
+			return err
+		}
+
+		// Fetch policies with the user context. Return any error that's not sac.ErrResourceAccessDenied.
+		policiesVisibleToUser, err := d.policyStore.GetAllPolicies(ctx)
+		if err != nil && !errors.Is(err, sac.ErrResourceAccessDenied) {
+			return errors.Wrap(err, "retrieving all policies")
+		}
+
+		listOfPolicies := strings.Join(intersectPolicies(policiesVisibleToUser, policiesContainingID), ",")
+
+		return errox.ReferencedByAnotherObject.Newf("cannot delete signature integration %q since there are "+
+			"existing policies that reference it: [%s]", integration.GetName(), listOfPolicies)
+	}
+
+	return nil
+}
+
+func checkIfPolicyContainsID(id string, policy *storage.Policy) bool {
+	for _, section := range policy.GetPolicySections() {
+		for _, group := range section.GetPolicyGroups() {
+			// Only check values of the "Image Signature Verified By" field.
+			if group.GetFieldName() == search.ImageSignatureVerifiedBy.String() {
+				for _, v := range group.GetValues() {
+					if v.GetValue() == id {
+						return true
+					}
+				}
+			}
+		}
+	}
+	return false
+}
+
+// intersectPolicies will intersect the policy names of policies which are visible to a user
+// and policies which have references of signature integrations.
+// It will return a list of policy names that are visible to the user, replacing policies
+// that have a reference but are not visibile to the user with "<hidden>".
+// Note: It is expected that policiesWithReferences is non-empty.
```

Comment:
What will happen if the list is empty?

Reply:
Ah, that's left-over from a previous commit where this was required, now it's not required anymore. It will just return an empty list to print, I'll remove the comment.